src/qnx/qnx-host-session-glue.c

```diff
@@ -0,0 +1,279 @@
+#include "frida-core.h"
+
+#include <gum/gum.h>
+#include <errno.h>
+#include <sys/elf.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#ifdef HAVE_SYS_USER_H
+# include <sys/user.h>
+#endif
+
+#define FRIDA_OFFSET_E_ENTRY 0x18
+#if defined (HAVE_I386)
+# define FRIDA_SIGBKPT SIGTRAP
+#elif defined (HAVE_ARM)
+# define FRIDA_SIGBKPT SIGBUS
+#endif
+
+#define CHECK_OS_RESULT(n1, cmp, n2, op) \
+  if (!(n1 cmp n2)) \
+  { \
+    failed_operation = op; \
+    goto handle_os_error; \
+  }
+
+#if defined (HAVE_I386)
+# define regs_t struct user_regs_struct
+#elif defined (HAVE_ARM)
+# define regs_t struct pt_regs
+#else
+# error Unsupported architecture
+#endif
+
+typedef struct _FridaSpawnInstance FridaSpawnInstance;
+typedef struct _FridaProbeElfContext FridaProbeElfContext;
+
+struct _FridaSpawnInstance
+{
+  FridaQnxHostSession * host_session;
+  pid_t pid;
+};
+
+struct _FridaProbeElfContext
+{
+  pid_t pid;
+  GumAddress entry_point;
+  gsize word_size;
+};
+
+static FridaSpawnInstance * frida_spawn_instance_new (FridaQnxHostSession * host_session);
+static void frida_spawn_instance_free (FridaSpawnInstance * instance);
+static void frida_spawn_instance_resume (FridaSpawnInstance * self);
+
+static gboolean frida_wait_for_child_signal (pid_t pid, int signal);
+static gboolean frida_run_to_entry_point (pid_t pid, GError ** error);
+
+static gboolean frida_examine_range_for_elf_header (const GumRangeDetails * details, gpointer user_data);
+
+guint
+_frida_qnx_host_session_do_spawn (FridaQnxHostSession * self, const gchar * path, gchar ** argv, int argv_length, gchar ** envp, int envp_length, GError ** error)
+{
+  FridaSpawnInstance * instance;
+  int status;
+  //long ret;
+  gboolean success;
+  const gchar * failed_operation;
+
+  instance = frida_spawn_instance_new (self);
+
+  g_assert_not_reached ();
+
+  instance->pid = fork ();
+  if (instance->pid == 0)
+  {
+   // ptrace (PTRACE_TRACEME, 0, NULL, NULL);
```

Comment:
Note that I have not implemented spawn at all (partially because there are no test cases for spawning and I'm too lazy to implement test cases). I've commented out all the ptrace calls so that the qnx-host-session code compiles. I'm going to leave them with this comment style for now until we actually implement spawning and all these comments will disappear.